agent/consul/leader_connect_ca.go

```diff
@@ -154,16 +154,24 @@ func (c *CAManager) setState(newState caState, validateState bool) (caState, err
 	state := c.state
 
 	if !validateState ||
-		state == caStateInitialized ||
+		(state == caStateInitialized && newState != caStateInitializing) ||
 		(state == caStateUninitialized && newState == caStateInitializing) ||
 		(state == caStateUninitialized && newState == caStateReconfig) {
 		c.state = newState
 	} else {
-		return state, fmt.Errorf("CA is already in state %q", state)
+		return state, &caStateError{Current: state}
```

Comment:
returning the state 2 times here is confusing. The only reason we return the state as part of `SetState` is to check if it changed, if we decide that we will return an error if the state is not changing or if the state transition is illegal and that error have the state embedded, we should not return the state anymore, only an error struct. If the error is nil that mean that the `SetState` succeeded and the new state is the one we passed as parameter.